File: typescript/medium/RangeFreqQuery.ts
class RangeFreqQuery {
  private map: Map<number, number[]>;
  constructor(arr: number[]) {
    this.map = new Map<number, number[]>();
    for (let i = 0; i < arr.length; i++) {
      if (!this.map.has(arr[i])) {
        this.map.set(arr[i], new Array<number>());
      }
      this.map.get(arr[i])?.push(i);
    }
  }

  query(left: number, right: number, value: number): number {
    if (!this.map.has(value)) {
      return 0;
    }
    const arr = this.map.get(value)!;
    const l = this.lower_bound(arr, left);
    const r = this.upper_bound(arr, right);
    return r - l;
  }

  lower_bound(arr: number[], target: number) {
    let left = 0;
    let right = arr.length - 1;
    while (left <= right) {
      const mid = left + Math.floor((right - left) / 2);
      if (arr[mid] < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    return left;
  }

  upper_bound(arr: number[], target: number): number {
    let left = 0;
    let right = arr.length - 1;
    while (left <= right) {
      const mid = left + Math.floor((right - left) / 2);
      if (arr[mid] <= target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    return left;
  }
}

File: java/src/medium/RangeFreqQuery.java
package medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class RangeFreqQuery {
    private HashMap<Integer, List<Integer>> map;

    public RangeFreqQuery(int[] arr) {
        map = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            if (!map.containsKey(arr[i])) {
                map.put(arr[i], new ArrayList<>());
            }
            map.get(arr[i]).add(i);
        }
    }

    public int query(int left, int right, int value) {
        if (!map.containsKey(value)) {
            return 0;
        }
        List<Integer> arr = map.get(value);
        int l = -1;
        int start = 0;
        int end = arr.size() - 1;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (arr.get(mid) < left) {
                l = mid;
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        int r = arr.size();
        start = 0;
        end = arr.size() - 1;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (arr.get(mid) > right) {
                r = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return r - l;
    }

    public static void main(String[] args) {
        int[] arr = {1, 1, 1, 2, 2};
        RangeFreqQuery rangeFreqQuery = new RangeFreqQuery(arr);
        rangeFreqQuery.query(0, 1, 2);
    }
}


